### ISSUE TITLE
[FEATURE REQUEST] PXE Boot over IB

### ISSUE THREAD

**Issue by strus38** [bluebanquise/bluebanquise#825]
Support of PXE boot over IB for servers not providing an Ethernet card by default.

## Context
Some servers do not provide a default ethernet card, so we need to be able to boot over IB (diskful/diskless)

## Feature
Being able to boot pxe without Ethernet card.

## Additional data
Seems like technically feasible, just make it available in BB.

**Comment by oxedions** [comment]
Dear @strus38 

Thank you for this feature request. 🙂 

I had a quick look. According to Nvidia/Mellanox documentation, the following elements are required:

1. Dhcp with client-identifier available (https://docs.nvidia.com/networking/pages/viewpage.action?pageId=39284368). This is the case in BB 2.0 collections, so that part is covered.
2. iPXE rom mellanox drivers. BB roms should already embed these drivers. I recommend using SNP - ONLY roms.
3. Additional kernel parameters, this is covered since BB 1.X and also in BB 2.0, via ep_ dedicated variable.
4. Specific modules into initramfs. In theory, according to documentation, OFED drivers should be present in RHEL 9 iso initramfs (to be confirmed). If not, it will be required to extract initramfs, add MOFED or OFED drivers, and repack it.

So in theory, the collections already allow such usage. I however cannot confirm since I do not have such hardware at disposal.
I will check what I can do on my side.

**Comment by capitn198** [comment]
Hello,
Here the results of my tests.

Context:
  - RHEL 8.7
  - bluebanquise 1.5.2
  - Hardware:
    - Motherboard: Gigabyte H263-S63
    - CPU: 2x Intel(R) Xeon(R) Gold 5418Y
    - 1x dedicated Ethernet port for BMC (not for OS)
    - IB HBA: 1x Mellanox Technologies MT2892 Family [ConnectX-6 Dx] HDR
    - 1x IB switch MQM8790-HS2X_Ax unmanaged

### **PXE diskfull** installation over IB link is working.
Need some tweak in the configuration of the inventory.

In the node definition (ex: /etc/bluebanquise/inventory/cluster/compute.yml):
```
mg_compute:
  children:
    equipment_compute:
      hosts:
        n01:
          bmc:
            name: n01-bmc
            ip4: 192.168.170.1
            mac: 74:56:3c:5b:93:e5
            network: admin1-2
          network_interfaces:
            - interface: ib0
              ip4: 192.168.169.1
#$ flint -d lid-5 q|grep Base
#Base GUID:             1070fd0300a24664        4
#Base MAC:              1070fda24664            4
              dhcp_client_identifier: 20:10:70:fd:03:00:a2:46:64 # Request from OS
              mac: 10:70:fd:a2:46:64 # Request from PXE
              network: admin1-1
              type: infiniband
``` 
Note that the MAC address used during PXE is 10:70:fd:a2:46:64 and during OS startup is 20:10:70:fd:03:00:a2:46:64.
That's why we have 2x MAC addresses for the same interface.
As mentioned in the comment, the `flint` command can be used to get these MAC addresses (add the "20:" prefix for the OS MAC in my situation).

We need also to adjust the equipment definition like:
```
$ head /etc/bluebanquise/inventory/group_vars/equipment_compute/equipment_profile.yml
ep_ipxe_platform: efi    # can be pcbios (for legacy/bios) or efi
ep_ipxe_driver: snponly                   # default or snp or snponly
ep_kernel_parameters: nomodeset bootdev=ib0 ksdevice=ib0 net.ifnames=0 biosdevname=0 rd.neednet=1 rd.bootif=0 rd.driver.pre=mlx5_ib,mlx4_ib,ib_ipoib ip=ib0:dhcp rd.net.dhcp.retry=10 rd.net.timeout.iflink=60 rd.net.timeout.ifup=80 rd.net.timeout.carrier=80
``` 

With this configuration it is possible to install the node through the IB link and then apply configuration with ansible-playbook.

### PXE diskless installation over IB link is working
The same changes made in the previous section (diskfull) must be done.

With diskless  i was obliged to generate the initramfs manually otherwise the node won't boot.
The initrd generated by `disklessset` was not working.

Once the image is mounted with the `disklessset` utility, I made (the name of my image is "test"):

```
$ chroot /var/tmp/diskless/workdir/test/mnt/
  # Tell the OS to do DHCP on ib0 at boot time
  # can be done in an ansible role
  $ cat /etc/sysconfig/network-scripts/ifcfg-ib0
CONNECTED_MODE=no
TYPE=InfiniBand
PROXY_METHOD=none
BROWSER_ONLY=no
BOOTPROTO=dhcp
DEFROUTE=yes
IPV4_FAILURE_FATAL=no
IPV6INIT=yes
IPV6_AUTOCONF=yes
IPV6_DEFROUTE=yes
IPV6_FAILURE_FATAL=no
IPV6_ADDR_GEN_MODE=eui64
NAME=ib0
DEVICE=ib0
ONBOOT=yes

  # Generate initramfs with the IB drivers inside
  $ dnf install dracut-live
  $ echo 'install mlx5_core /sbin/modprobe --ignore-install mlx5_core; /sbin/modprobe mlx5_ib; /sbin/modprobe ib_ipoib' >> /etc/modprobe.d/mlx.conf
echo 'add_drivers+="mlx5_ib ib_ipoib"' > /etc/dracut.conf.d/mlx.conf
  $ /usr/bin/dracut --xz -v -a network -a base -a nfs --force-add livenet --add-drivers xfs --no-hostonly --nolvmconf -f
  $ exit

# Manually copy the generated initramfs in the place where `disklessset` put it
$ cp /var/tmp/diskless/workdir/test/mnt/boot/initramfs-4.18.0-425.19.2.el8_7.x86_64.img /var/www/html/preboot_execution_environment/diskless/kernels/initramfs-kernel-4.18.0-425.19.2.el8_7.x86_64
``` 

With this initramfs, the diskless OS boots fine.

**Comment by oxedions** [comment]
Dear @capitn198 

Many thanks for this feedback, and my apologies for the delay.
This is very interesting ! So it could work out of the box with only documentation.

I will add this content to main documentation asap. I will let this issue open until its done 😉

**Comment by capitn198** [comment]
**UPDATE**

I don't know exactly what happened but the PXE boot over IB is not working anymore on my hardware.
The DHCP server sees the DHCPDISCOVER from the IB HBA and replies with the DHCPOFFER but the iPXE ROM never sends the DHCPREQUEST/DHCPACK.
My guess is that the firmware upgrade of the HBA broke the PXE boot but I was not able to repair with a downgrade to the original version (maybe a config in the NVRAM of the HBA -> mlxconfig).

Nevertheless I succeeded to PXE boot the nodes over IB using the iPXE ROM https://boot.ipxe.org/ipxe.iso using the BMC Vitual Media feature.
So the BMC emulates a CDROM with the iPXE ROM and it makes DHCP requests using the IB HBA.
It is possible to do this from the KVM provided by the BMC (HTTP interface).

Gigabyte provides a tool named `gbtutility` to interact with the BMC in command line.
I attach a little script I made that show how to mount an ISO file accessible from a HTTP server by the BMC.

[gbtutility-mount-ipxe-cdrom.sh.gz](https://github.com/bluebanquise/bluebanquise/files/12741847/gbtutility-mount-ipxe-cdrom.sh.gz)
